src/deploy/functions/runtimes/golang/index.ts

```diff
@@ -0,0 +1,162 @@
+  const sourceDir = options.config.path(sourceDirName);
+  const goModPath = path.join(sourceDir, "go.mod");
+  const projectId = getProjectId(options);
+
+  let module: Module;
+  try {
+    const modBuffer = await promisify(fs.readFile)(goModPath);
+    module = parseModule(modBuffer.toString("utf8"));
+  } catch (err) {
+    logger.debug("Customer code is not Golang code (or they aren't using modules)");
+    return;
+  }
+
+  let runtime = options.config.get("functions.runtime");
+  if (!runtime) {
+    if (!module.version) {
+      throw new FirebaseError("Could not detect Golang version from go.mod");
+    }
+    if (!VERSION_TO_RUNTIME[module.version]) {
+      throw new FirebaseError(
+        `go.mod specifies Golang version ${
+          module.version
+        } which is unsupported by Google Cloud Functions. Valid values are ${Object.keys(
+          VERSION_TO_RUNTIME
+        ).join(", ")}`
+      );
+    }
+    runtime = VERSION_TO_RUNTIME[module.version];
+  }
+
+  return new Delegate(projectId, sourceDirName, sourceDir, runtime, module);
+}
+
+// A module can be much more complicated than this, but this is all we need so far.
+// for a full reference, see https://golang.org/doc/modules/gomod-ref
+interface Module {
+  module: string;
+  version: string;
+  dependencies: Record<string, string>;
+}
+
+export function parseModule(mod: string): Module {
+  const module: Module = {
+    module: "",
+    version: "",
+    dependencies: {},
+  };
+  const lines = mod.split("\n");
+  let inRequire = false;
+  for (const line of lines) {
+    if (inRequire) {
+      const endRequireMatch = /\)/.exec(line);
+      if (endRequireMatch) {
+        inRequire = false;
+        continue;
+      }
+
+      const requireMatch = /([^ ]+) (.*)/.exec(line);
+      if (requireMatch) {
+        module.dependencies[requireMatch[1]] = requireMatch[2];
+        continue;
+      }
+
+      if (line.trim()) {
+        logger.debug("Don't know how to handle line", line, "inside a mod.go require block");
+      }
+      continue;
+    }
+    const modMatch = /^module (.*)$/.exec(line);
+    if (modMatch) {
+      module.module = modMatch[1];
+      continue;
+    }
+    const versionMatch = /^go (\d+\.\d+)$/.exec(line);
+    if (versionMatch) {
+      module.version = versionMatch[1];
+      continue;
+    }
+
+    const requireMatch = /^require ([^ ]+) (.*)$/.exec(line);
+    if (requireMatch) {
+      module.dependencies[requireMatch[1]] = requireMatch[2];
+      continue;
+    }
+
+    const requireBlockMatch = /^require +\(/.exec(line);
+    if (requireBlockMatch) {
+      inRequire = true;
+      continue;
+    }
+
+    if (line.trim()) {
+      logger.debug("Don't know how to handle line", line, "in mod.go");
+    }
+  }
+
+  return module;
+}
+
+export class Delegate {
+  public readonly name = "golang";
+
+  constructor(
+    private readonly projectId: string,
+    private readonly sourceDirName: string,
```

Comment:
I don't love that we have both sourceDirName and sourceDir - its a bit unclear at first what each means, it seems like you can always infer sourceDirName from sourceDir.

Reply:
There are three strings: project root, sourceDir, and sourceDirName. SourceDirName will almost always be "functions" but it could soon be "functions/golang" or similar, so I can't just pop the top folder name from sourceDir. I'll rename sourceDirName though. I agree it's confusing but it came from prior code.